juju/asyncprovisioner.py

```diff
@@ -0,0 +1,260 @@
+import os
+import re
+import shlex
+import tempfile
+import uuid
+from subprocess import CalledProcessError
+
+import asyncio
+import paramiko
```

Comment:
You're still importing this even though you removed it from the dependency list.

Reply:
@johnsca And if for simplicity I just replace the current one with the new one?

Reply:
I would much prefer to just drop the sync version entirely but it was @SimonRichardson's concern that it breaks an API that was potentially being used, though I strongly suspect that it was only ever used via the model method anyway, rather than being used directly. I lean toward just saying that it was an internal implementation detail and not worrying about the major version bump, but I do understand the concerns.

Thoughts, @SimonRichardson @ChrisMacNaughton?

Reply:
Honestly, I'd be tempted to minimally wrap the Async version into a Sync version that maintains the current API and logs a warning about the deprecation now, and then remove it entirely soonish

Reply:
I agree with @ChrisMacNaughton on this one and by soonish, another major version of juju 2.9.

Reply:
Unfortunately, calling async code from sync code can only be done in asyncio if no other async code is running (that is, there is no existing event loop).  It seems likely that that if the existing sync implementation class has been used directly, it was likely from async code as in the current implementation of `add_machine`.  That means if we don't want to break it, we have to wrap in the other direction using `run_in_executor` and a thread, refactor the existing class to also use subprocess, or just leave the existing one alone and keep the paramiko dependency until we remove the deprecated version.